### PR TITLE
update url_shorner model,serializer and views

### PR DESCRIPTION
**Pull request by osadolortech**

### CODE REVIEW

File: shortapi/models.py [hunk 1]
```diff
@@ -1,3 +1,28 @@
...
+class Link(models.Model):
+    user = models.ForeignKey(User,related_name="user",on_delete=models.CASCADE)
+    original_url = models.URLField()
+    shorten_url = models.URLField(blank=True, null=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    def shortener(self):
+        while True:
+            random_string = ''.join(choices(ascii_letters,k=6))
+            new_link = settings.HOST_URL+'/'+random_string
```
Comment (iamr0b0tx) [inline shortapi/models.py]: nice work using HOST_URL in settings

File: shortapi/models.py [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+            if not Link.objects.filter(shorten_url=new_link).exists():
+                break
+        return new_link
+
+    def save(self,*args,**kwargs):
+        if not self.shorten_url:
+            new_link = self.shortener()
+            self.shorten_url=new_link
+        return super().save(*args,**kwargs)
```
Comment (iamr0b0tx) [inline shortapi/models.py]: use a django pre_save signal instead

File: user/urls.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from rest_framework.urls import path
+from .views import Register
+
+
+
+urlpatterns =[
+    path("", Register.as_view(), name=('register'))
```
Comment (iamr0b0tx) [inline user/urls.py]: this url file is not needed. just use the one in the project dir

File: user/views.py [hunk 4]
```diff
@@ -1,3 +1,15 @@
...
 # Create your views here.
+
+
+class Register(CreateAPIView):
+    serializer_class = RegisterSerializer
+    def post(self,request):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        User.objects.create_user(**serializer.validated_data)
+        return Response({"user: created user"})
```
Comment (iamr0b0tx) [inline user/views.py]: I already told you to use auth frameworks

File: user/serializers.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+from rest_framework import serializers
+from .models import User
+from django.contrib.auth.password_validation import validate_password
+from rest_framework.validators import UniqueValidator
+
+class RegisterSerializer(serializers.Serializer):
+    name = serializers.CharField()
+    email = serializers.EmailField(required=True,write_only=True,validators=[UniqueValidator(queryset=User.objects.all())])
+    password = serializers.CharField(required= True,write_only=True, validators=[validate_password])
+    password2 = serializers.CharField(required= True,write_only=True)
```
Comment (iamr0b0tx) [inline user/serializers.py]: use auth frameworks I recommended

File: user/models.py [hunk 6]
```diff
@@ -18,18 +19,20 @@ def create_superuser(self,email,password,**extra_fields):
...
     email = models.EmailField(unique=True)
     name = models.CharField(max_length=125)
     created_at =models.DateTimeField(auto_now_add=True)
-    update_at = models.DateTimeField(auto_now=True)
+    updated_at = models.DateTimeField(auto_now=True)
     is_staff = models.BooleanField(default=False)
     is_superuser = models.BooleanField(default=False)
-    is_staff = models.BooleanField(default=True)
+    is_active= models.BooleanField(default=True)
+
```
Comment (iamr0b0tx) [inline user/models.py]: why not just use AbstractUser class?

File: user/models.py [hunk 7]
```diff
@@ -18,18 +19,20 @@ def create_superuser(self,email,password,**extra_fields):
         extra_fields.setdefault('is_superuser',True)
         extra_fields.setdefault('is_active',True)
         extra_fields.setdefault('name','admin')
-        if extra_fields.get('is_staff'):
+        extra_fields.setdefault('password2', None)
+        if extra_fields.get('is_staff') is not True:
             raise ValueError("superuser must have is superuser as True")
         return self.create_user(email,password,**extra_fields)
```
Comment (iamr0b0tx) [inline user/models.py]: this may not be neccessary when using abstractUser class. I am talking about the custom manager

File: shortapi/views.py [hunk 8]
```diff
@@ -1,3 +1,22 @@
...
+from .serializers import LinkSerializer
+from .models import Link
+
+
+class ShortenerView(viewsets.ModelViewSet):
+    queryset = Link.objects.all()
+    serializer_class = LinkSerializer
+
+    def perform_create(self, serializer):
+        return super().perform_create(serializer)
```
Comment (iamr0b0tx) [inline shortapi/views.py]: is this even neccessary?

File: shortapi/models.py [hunk 9]
```diff
@@ -1,3 +1,28 @@
...
+from string import ascii_letters
+
+
+
+class Link(models.Model):
+    user = models.ForeignKey(User,related_name="user",on_delete=models.CASCADE)
+    original_url = models.URLField()
+    shorten_url = models.URLField(blank=True, null=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+
```
Comment (iamr0b0tx) [inline shortapi/models.py:15]: updated_at may be a good addition

File: shortapi/models.py [hunk 10]
```diff
@@ -1,3 +1,28 @@
...
+    def shortener(self):
```
Comment (iamr0b0tx) [inline shortapi/models.py]: separate this into a new file as a utility function or service that is without side effects i.e. no db calls

**Review by iamr0b0tx — CHANGES_REQUESTED** [review]

File: user/models.py (hunk 7 above)
Comment (osadolortech) [reply user/models.py]: ok sir

File: shortapi/views.py (hunk 8 above)
Comment (osadolortech) [reply shortapi/views.py]: not exactly sir

File: user/models.py (hunk 7 above)
Comment (iamr0b0tx) [reply user/models.py]: did you confirm this?

File: shortapi/views.py (hunk 8 above)
Comment (iamr0b0tx) [reply shortapi/views.py]: so why have it?

File: shortapi/models.py (hunk 9 above)
Comment (iamr0b0tx) [reply shortapi/models.py:15]: why not?

Comment (osadolortech) [reply shortapi/models.py:15]: this was a mistake sir already added the update_at

Comment (osadolortech) [reply shortapi/models.py:15]: this was a mistake sir already added the update_at

File: shortapi/views.py (hunk 8 above)
Comment (osadolortech) [reply shortapi/views.py]: remove it already sir, i wanted to overwrite the perform_create method but i realized i didnt need to any more so i just didnt remove it but is removed now

Comment (osadolortech) [reply shortapi/views.py]: remove it already sir, i wanted to overwrite the perform_create method but i realized i didnt need to any more so i just didnt remove it but is removed now

File: user/models.py (hunk 7 above)
Comment (osadolortech) [reply user/models.py]: yes sir removed the ustomanger

Comment (osadolortech) [reply user/models.py]: yes sir removed the ustomanger

File: user/models.py (hunk 6 above)
Comment (osadolortech) [reply user/models.py]: using Abstract user now sir

File: user/serializers.py (hunk 5 above)
Comment (osadolortech) [reply user/serializers.py]: changes implemented

File: user/views.py (hunk 4 above)
Comment (osadolortech) [reply user/views.py]: changes implemented

File: shortapi/models.py (hunk 2 above)
Comment (osadolortech) [reply shortapi/models.py]: done sir

File: shortapi/models.py (hunk 1 above)
Comment (osadolortech) [reply shortapi/models.py]: thank you sir

File: shortapi/views.py (hunk 8 above)
Comment (iamr0b0tx) [reply shortapi/views.py]: then why is it still here?

**Review by iamr0b0tx — CHANGES_REQUESTED** [review]

File: shortapi/models.py (hunk 9 above)
Comment (iamr0b0tx) [reply shortapi/models.py:15]: I don't see it

**Comment by osadolortech** [comment]
Oh sorry sir I just pushed the changes

**Review by iamr0b0tx — APPROVED** [review]

**Pull request merged**